File: app/src/main/java/com/hqq/example/ui/customize/widget/TemplatesImage.java
package com.hqq.example.ui.customize.widget;

import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Matrix;
import android.graphics.Paint;
import android.graphics.Path;
import android.graphics.drawable.Drawable;
import android.util.AttributeSet;
import android.view.MotionEvent;
import android.view.View;

import androidx.annotation.Nullable;
import androidx.core.content.ContextCompat;

import com.hqq.core.utils.log.LogUtils;
import com.hqq.example.R;
import com.hqq.example.ui.customize.widget.imageedit.DrawableSticker;

import java.util.Arrays;

/**
 * @Author : huangqiqiang
 * @Package : com.hqq.example.ui.customize.widget
 * @FileName :   TemplatesImage
 * @Date : 2020/1/10 0010  下午 1:35
 * @Email : qiqiang213@gmail.com
 * @Descrive :
 */
public class TemplatesImage extends View {
    public TemplatesImage(Context context) {
        super(context);
        init();
    }

    public TemplatesImage(Context context, @Nullable AttributeSet attrs) {
        super(context, attrs);
        init();

    }

    public TemplatesImage(Context context, @Nullable AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        init();
    }

    Path mClipPath = new Path();
    float distanceX = 0;
    float distanceY = 0;
    float mKeyX;
    float mKeyY;
    float mFrame = getWidth() / 8;
    Matrix mCurrentMatrix;
    private Paint borderPaint;

    private void init() {
        mCurrentMatrix = new Matrix();
        mClipPath.moveTo(0, 0);
        mClipPath.lineTo(300, 0);
        mClipPath.lineTo(300, 300);
        mClipPath.lineTo(0, 300);
        mClipPath.lineTo(0, 0);
        mDrawableSticker = new DrawableSticker(ContextCompat.getDrawable(getContext(), R.mipmap.ic_content));
    }

    @Override
    protected void onLayout(boolean changed, int left, int top, int right, int bottom) {
        super.onLayout(changed, left, top, right, bottom);
        mFrame = getWidth() / 8;
        LogUtils.e("onLayout", "" + getWidth());
    }

    DrawableSticker mDrawableSticker;

    @Override
    protected void onDraw(Canvas canvas) {
        super.onDraw(canvas);
        initClipPath();

        drawBg(canvas);


        drawClipBg(canvas);

        mDrawableSticker.draw(canvas);

        float[] dst = new float[8];
        mDrawableSticker.getMatrix()
                .mapPoints(dst, getBitmapSize());
        LogUtils.e(Arrays.toString(dst));
        float x1 = dst[0];
        float y1 = dst[1];
        float x2 = dst[2];
        float y2 = dst[3];
        float x3 = dst[4];
        float y3 = dst[5];
        float x4 = dst[6];
        float y4 = dst[7];
        borderPaint = new Paint();
        borderPaint.setAntiAlias(true);
        borderPaint.setColor(Color.RED);
//        borderPaint.setAlpha(128);
        borderPaint.setStrokeWidth(6);
        canvas.drawLine(x1, y1, x2, y2, borderPaint);
        canvas.drawLine(x1, y1, x3, y3, borderPaint);
        canvas.drawLine(x2, y2, x4, y4, borderPaint);
        canvas.drawLine(x4, y4, x3, y3, borderPaint);


    }

    private float[] getBitmapSize() {
        return new float[]{
                0f, 0f, mDrawableSticker
                .getWidth(), 0f, 0f, mDrawableSticker.getHeight(), mDrawableSticker.getWidth(), mDrawableSticker.getHeight()
        };
    }

    boolean isTounch = false;

    @Override
    public boolean dispatchTouchEvent(MotionEvent event) {
        switch (event.getAction()) {
            case MotionEvent.ACTION_DOWN:
                LogUtils.e("getRight", getRight());
                LogUtils.e("ACTION_DOWN -----------" + event.getX());
                mKeyX = event.getX();
                mKeyY = event.getY();

                if (mDrawableSticker.contains(mKeyX, mKeyY)) {
                    isTounch = true;

                    mCurrentMatrix.set(mDrawableSticker.getMatrix()
                    );
                }

                break;
            case MotionEvent.ACTION_MOVE:
                LogUtils.e("ACTION_MOVE -----------" + event.getX());
                distanceX = event.getX() - mKeyX;
                distanceY = event.getY() - mKeyY;
//                mKeyX = event.getX();
//                mKeyY = event.getY();


                float xx = ((getWidth() - mFrame * 2) / getWidth());
                float yy = ((getHeight() - mFrame * 2) / getHeight());


                LogUtils.e("ACTION_MOVE --------distanceX---" + distanceX);
                LogUtils.e("ACTION_MOVE --------distanceY---" + distanceY);
                LogUtils.e("-------------------     " + xx + "            " + yy);

                if (isTounch) {

                    mDrawableSticker.getMatrix()
                            .set(mCurrentMatrix);

                    mDrawableSticker.getMatrix()
                            .postTranslate(distanceX, distanceY);
                    invalidate();
                }


                break;
            case MotionEvent.ACTION_UP:
            case MotionEvent.ACTION_CANCEL:
                isTounch = false;
                LogUtils.e("---------------------------------------------------------------------------------------------------------------------");
                LogUtils.e("distanceX", distanceX);
                LogUtils.e("getRawX -----------" + event.getRawX());
                LogUtils.e("getX", getX());
                LogUtils.e("getLeft", getLeft());
                LogUtils.e("getRight", getRight());
                LogUtils.e("getTranslationX", getTranslationX());
                LogUtils.e("getScrollX", getScrollX());

                break;
            default:

        }


        return super.dispatchTouchEvent(event);
    }


    @Override
    public boolean onTouchEvent(MotionEvent event) {
        switch (event.getAction()) {
            case MotionEvent.ACTION_DOWN:
                return true;
            default:
        }
        return super.onTouchEvent(event);
    }


    private void initClipPath() {
        mClipPath.reset();
        mClipPath.moveTo(mFrame, mFrame);
        mClipPath.lineTo(getWidth() - mFrame, mFrame);
        mClipPath.lineTo(getWidth() - mFrame, getHeight() - mFrame);
        mClipPath.lineTo(mFrame, getHeight() - mFrame);
        mClipPath.lineTo(mFrame, mFrame);
    }

    private void drawClipBg(Canvas canvas) {

        Paint paint = new Paint();
        paint.setColor(Color.RED);
        paint.setStrokeWidth(5);
        paint.setStyle(Paint.Style.STROKE);
        canvas.drawPath(mClipPath, paint);
        canvas.clipPath(mClipPath);
        refreshDrawableState();
    }

    private void drawBg(Canvas canvas) {
        canvas.save();
        Bitmap bitmap = BitmapFactory.decodeResource(getResources(), R.mipmap.bg_sty);
        Matrix matrix = new Matrix();
        matrix.setScale(Float.valueOf(getWidth()) / bitmap.getWidth(), Float.valueOf(getHeight()) / bitmap.getHeight());
        canvas.drawBitmap(bitmap, matrix, null);
        canvas.restore();
    }
}

File: app/src/main/java/com/easy/example/ui/customize/widget/TemplatesImage.kt
package com.easy.example.ui.customize.widget

import android.content.Context
import android.graphics.*
import android.util.AttributeSet
import android.view.MotionEvent
import android.view.View
import androidx.core.content.ContextCompat
import com.easy.core.utils.log.LogUtils.dInfo
import com.easy.core.utils.log.LogUtils.e
import com.easy.example.R
import com.easy.example.ui.customize.widget.imageedit.DrawableSticker
import java.util.*

/**
 * @Author : huangqiqiang
 * @Package : com.easy.example.ui.customize.widget
 * @FileName :   TemplatesImage
 * @Date : 2020/1/10 0010  下午 1:35
 * @Email : qiqiang213@gmail.com
 * @Descrive :
 */
class TemplatesImage : View {
    constructor(context: Context?) : super(context) {
        init()
    }

    constructor(context: Context?, attrs: AttributeSet?) : super(context, attrs) {
        init()
    }

    constructor(context: Context?, attrs: AttributeSet?, defStyleAttr: Int) : super(context, attrs, defStyleAttr) {
        init()
    }

    var mClipPath = Path()
    var distanceX = 0f
    var distanceY = 0f
    var mKeyX = 0f
    var mKeyY = 0f
    var mFrame = width / 8.toFloat()
    var mCurrentMatrix: Matrix? = null
    private var borderPaint: Paint? = null
    private fun init() {
        mCurrentMatrix = Matrix()
        mClipPath.moveTo(0f, 0f)
        mClipPath.lineTo(300f, 0f)
        mClipPath.lineTo(300f, 300f)
        mClipPath.lineTo(0f, 300f)
        mClipPath.lineTo(0f, 0f)
        mDrawableSticker =
            com.easy.example.ui.customize.widget.imageedit.DrawableSticker(ContextCompat.getDrawable(context, R.mipmap.ic_content))
    }

    override fun onLayout(changed: Boolean, left: Int, top: Int, right: Int, bottom: Int) {
        super.onLayout(changed, left, top, right, bottom)
        mFrame = width / 8.toFloat()
        dInfo("onLayout" + width)
    }

    var mDrawableSticker: com.easy.example.ui.customize.widget.imageedit.DrawableSticker? = null
    override fun onDraw(canvas: Canvas) {
        super.onDraw(canvas)
        initClipPath()
        drawBg(canvas)
        drawClipBg(canvas)
        mDrawableSticker!!.draw(canvas)
        val dst = FloatArray(8)
        mDrawableSticker!!.matrix
                .mapPoints(dst, bitmapSize)
        dInfo(Arrays.toString(dst))
        val x1 = dst[0]
        val y1 = dst[1]
        val x2 = dst[2]
        val y2 = dst[3]
        val x3 = dst[4]
        val y3 = dst[5]
        val x4 = dst[6]
        val y4 = dst[7]
        borderPaint = Paint()
        borderPaint!!.isAntiAlias = true
        borderPaint!!.color = Color.RED
        //        borderPaint.setAlpha(128);
        borderPaint!!.strokeWidth = 6f
        canvas.drawLine(x1, y1, x2, y2, borderPaint!!)
        canvas.drawLine(x1, y1, x3, y3, borderPaint!!)
        canvas.drawLine(x2, y2, x4, y4, borderPaint!!)
        canvas.drawLine(x4, y4, x3, y3, borderPaint!!)
    }

    private val bitmapSize: FloatArray
        private get() = floatArrayOf(
                0f, 0f, mDrawableSticker
                ?.getWidth()!!.toFloat(), 0f, 0f, mDrawableSticker!!.height.toFloat(), mDrawableSticker!!.width.toFloat(), mDrawableSticker!!.height
                .toFloat())

    var isTounch = false
    override fun dispatchTouchEvent(event: MotionEvent): Boolean {
        when (event.action) {
            MotionEvent.ACTION_DOWN -> {
                dInfo("getRight"+ right)
                dInfo("ACTION_DOWN -----------" + event.x)
                mKeyX = event.x
                mKeyY = event.y
                if (mDrawableSticker!!.contains(mKeyX, mKeyY)) {
                    isTounch = true
                    mCurrentMatrix!!.set(mDrawableSticker!!.matrix
                    )
                }
            }
            MotionEvent.ACTION_MOVE -> {
                dInfo("ACTION_MOVE -----------" + event.x)
                distanceX = event.x - mKeyX
                distanceY = event.y - mKeyY
                //                mKeyX = event.getX();
//                mKeyY = event.getY();
                val xx = (width - mFrame * 2) / width
                val yy = (height - mFrame * 2) / height
                dInfo("ACTION_MOVE --------distanceX---$distanceX")
                dInfo("ACTION_MOVE --------distanceY---$distanceY")
                dInfo("-------------------     $xx            $yy")
                if (isTounch) {
                    mDrawableSticker!!.matrix
                            .set(mCurrentMatrix)
                    mDrawableSticker!!.matrix
                            .postTranslate(distanceX, distanceY)
                    invalidate()
                }
            }
            MotionEvent.ACTION_UP, MotionEvent.ACTION_CANCEL -> {
                isTounch = false
                dInfo("---------------------------------------------------------------------------------------------------------------------")
                e("distanceX"+ distanceX)
                e("getRawX -----------" + event.rawX)
                e("getX"+ x)
                e("getLeft"+ left)
                e("getRight"+ right)
                e("getTranslationX"+ translationX)
                e("getScrollX"+ scrollX)
            }
            else -> {
            }
        }
        return super.dispatchTouchEvent(event)
    }

    override fun onTouchEvent(event: MotionEvent): Boolean {
        when (event.action) {
            MotionEvent.ACTION_DOWN -> return true
            else -> {
            }
        }
        return super.onTouchEvent(event)
    }

    private fun initClipPath() {
        mClipPath.reset()
        mClipPath.moveTo(mFrame, mFrame)
        mClipPath.lineTo(width - mFrame, mFrame)
        mClipPath.lineTo(width - mFrame, height - mFrame)
        mClipPath.lineTo(mFrame, height - mFrame)
        mClipPath.lineTo(mFrame, mFrame)
    }

    private fun drawClipBg(canvas: Canvas) {
        val paint = Paint()
        paint.color = Color.RED
        paint.strokeWidth = 5f
        paint.style = Paint.Style.STROKE
        canvas.drawPath(mClipPath, paint)
        canvas.clipPath(mClipPath)
        refreshDrawableState()
    }

    private fun drawBg(canvas: Canvas) {
        canvas.save()
        val bitmap = BitmapFactory.decodeResource(resources, R.mipmap.bg_sty)
        val matrix = Matrix()
        matrix.setScale(java.lang.Float.valueOf(width.toFloat()) / bitmap.width, java.lang.Float.valueOf(height.toFloat()) / bitmap.height)
        canvas.drawBitmap(bitmap, matrix, null)
        canvas.restore()
    }
}